pkg/collector/corechecks/ebpf/probe/ebpfcheck/probe.go

```diff
@@ -595,41 +595,110 @@ func (e *entryCountBuffers) tryEnsureSizeForFullBatch(referenceMap *ebpf.Map) bo
 func (e *entryCountBuffers) resetBuffers() {
 	e.keys = nil
+	e.firstBatchKeys.reset()
 	e.values = nil
 	e.cursor = nil
 }
 
-func inplaceHash(data []byte, offset int, length int, seed uint32) uint32 {
+// inplaceSet is a set that stores the hashes of keys. It's only used for the specific case of this entry count, where
+// we want to check if we got a restarted iteration.
+//
+// Context: when iterating eBPF maps, the kernel returns a cursor that indicates the next key to get. If the map is changed
+// in between calls and that "next key" disappears, the kernel just starts the iteration from the beginning. This means that
+// we could be infinitely restarting the iteration if the map is constantly changing. To avoid this, we keep track of the
+// entries we have seen in the first batch lookup call. If for any subsequent batch we get a repeated key, then we know the
+// kernel restarted the iteration.
+//
+// Now, considering that we want to reduce as much as possible the memory usage, and that we are dealing with keys of arbitrary
+// sizes (only known at runtime), we cannot just use a map[[]byte]struct{} to store the keys. Solutions like using a string for the keys
+// would work, but they require extra copies and allocations. So, instead, we store the hashes of the keys in a map and check against
+// that later. This is not perfect as there is a chance of collision between hashes (0.003% for 131072 keys). However, saving memory
+// is more important and this approach lets us have a set without any allocations or copies, everything is done in-place as much as possible.
+//
+// One thing to note is that, in the case of a hash collision, we would detect a restart (false positive). The problem is that the collision
+// would always happen no matter how many restarts we have, which means that if we have two colliding keys in a map we would never
+// be able to get the number of entries. To mitigate that, we add the number of restarts we have to the hash as a "seed", so that if
+// we get restarted we will get a different hash and we should be able to get the number of entries.
+type inplaceSet struct {
+	set     map[uint32]struct{}
+	keySize int
+	seed    uint32
+}
+
+func (s *inplaceSet) reset() {
+	s.set = nil
+}
+
+// prepare prepares the set to store the given number of entries. Maps in Go grow automatically and never shrink, so we don't need to reallocate
+// anything. This just ensures that the map is initialized and ready to be used and with a size hint to reduce reallocations.
+func (s *inplaceSet) prepare(hintNumEntries int) {
+	if s.set == nil {
+		s.set = make(map[uint32]struct{}, hintNumEntries)
+	}
+}
+
+// clear deletes all entries in the set
+func (s *inplaceSet) clear() {
+	for k := range s.set {
+		delete(s.set, k)
+	}
+}
+
+// hash calculates the hash of the key at the given offset in a certain buffer
+// This is a FNV hash function with 32 bits
+func (s *inplaceSet) hash(data []byte, offset int) uint32 {
```

Comment:
Should we be using `hash/fnv` package instead?

Reply:
If not, please add a comment in the code why

Reply:
Found a way to add the seed without allocations/conversions, changed it to use hash/fnv.